src/util/inotify.c

```diff
@@ -378,23 +378,61 @@ static int watch_ctx_destructor(void *memptr)
     return 0;
 }
 
+errno_t get_absolute_path(const char *relative_path, char *absolute_path)
+{
+    char *filename = NULL;
+    char path[PATH_MAX + 1];
+    errno_t ret;
+
+    filename = strrchr(relative_path, '/');
+    if (filename == NULL || filename == relative_path) {
+        ret = ENOENT;
+        goto done;
+    }
+
+    memcpy(path, relative_path, filename - relative_path);
+    path[filename - relative_path] = '\0';
+
+    if (realpath(path, absolute_path) == NULL) {
+        ret = ENOENT;
+        goto done;
+    }
+
+    memcpy(absolute_path + strlen(absolute_path), filename, strlen(filename)+1);
+    ret = EOK;
+
+done:
+    return ret;
+}
+
 static errno_t resolve_filename(struct snotify_ctx *snctx,
                                 const char *filename,
                                 char *resolved,
                                 size_t resolved_size)
 {
-    /* NOTE: The code below relies in the GNU extensions for realpath,
-     * which will store in 'resolved' the prefix of 'filename' that does
-     * not exists if realpath call fails and errno is set to ENOENT */
     if (realpath(filename, resolved) == NULL) {
```

Comment:
Wouldn't it be more "readable" to skip `realpath()` attempt and use `get_absolute_path()` directly?

Reply:
No. `filename` can point to an existing file or directory, in that case the fail path is not necessary.